contrib/raftexample/kvstore.go

```diff
@@ -77,3 +97,34 @@ func (s *kvstore) readCommits(commitC <-chan *string, errorC <-chan error) {
 		log.Fatal(err)
 	}
 }
+
+func (s *kvstore) createSnapshot() ([]byte, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var buf bytes.Buffer
+	if err := gob.NewEncoder(&buf).Encode(s.kvStore); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func (s *kvstore) loadSnapshot(rd io.Reader) error {
+	var store map[string]string
+	if err := gob.NewDecoder(rd).Decode(&store); err != nil {
+		return err
+	}
+	s.mu.Lock()
+	s.kvStore = store
+	s.mu.Unlock()
+	return nil
+}
+
+func (s *kvstore) writeSnapshot(w io.Writer) error {
+	data, err := s.createSnapshot()
```

Comment:
probably we can pass in the writer into s.createSnap. then we do not need the buffer inside createSnapshot